### PR TITLE
vendor: bump etcd v3.3.27

### LINKED ISSUE #31182: docker daemon locks files in /dev
**Description**

A running docker daemon will lock files in /dev, other tools that want to temporary take such looks will block forever or fail.

**Steps to reproduce the issue:**

```
# run with
#
# rm -rf subdir
# make -O >/dev/null
#
# add -j2, then you can' t even terminate it

FILE := a b

all: $(addprefix subdir/,$(FILE))

$(addprefix subdir/,$(FILE)):
	echo "$@"
	mkdir -p $(dir $@); touch $@
```
1. Have the dockerd daemon running
2. Run the above Makefile with `make -O >/dev/null`

**Describe the results you received:**

make finishing

**Describe the results you expected:**

make is blocked waiting for a lock on /dev/null indefinitely. Stopping the dockerd service will allow it to continue.

**Additional information you deem important (e.g. issue happens only occasionally):**

Issue first appeared on January 19th, using the docker reporsitory `deb https://apt.dockerproject.org/repo debian-stretch main`
[Bug report for debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=855336)

**Output of `docker version`:**

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba372
 Built:        Wed Feb  8 06:44:30 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba372
 Built:        Wed Feb  8 06:44:30 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 9
Server Version: 1.13.1
Storage Driver: devicemapper
 Pool Name: docker-8:2-12328161-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 6.814 GB
 Data Space Total: 107.4 GB
 Data Space Available: 100.6 GB
 Metadata Space Used: 5.927 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.142 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.137 (2016-11-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-1-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 15.62 GiB
Name: desbuild
ID: 24GS:G7V4:MEKN:PBVA:MED3:JEMB:E5KG:5JHI:2SH6:7IGU:DOYL:Y4KH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by kolyshkin**
fixes https://github.com/moby/moby/issues/31182

Bump etcd to v3.3.27, which includes https://github.com/etcd-io/etcd/pull/12552,
to fix https://github.com/moby/moby/issues/31182

Full diff: https://github.com/coreos/etcd/compare/v3.3.25...v3.3.27

~This is a draft, pending a 3.3.26 release (https://github.com/etcd-io/etcd/issues/12698)~

### CODE REVIEW

**Comment by joakim-tjernlund** [comment]
Ping?

**Comment by thaJeztah** [comment]
I think we're waiting for the etcd maintainers to tag a new release; https://github.com/etcd-io/etcd/pull/12552#issuecomment-764723724

**Comment by kolyshkin** [comment]
Updated to use the upstream repo as the PR is merged. No changes other than the line in vendor.conf

**Comment by joakim-tjernlund** [comment]
CI still broken?

File: vendor/github.com/coreos/etcd/client/json.go
Comment (thaJeztah) [inline vendor/github.com/coreos/etcd/client/json.go:19]: Looks like it adds two new dependencies (that are currently not vendored), which looks to cause the CI failure

Comment (kolyshkin) [reply vendor/github.com/coreos/etcd/client/json.go:19]: fixed

**Comment by kolyshkin** [comment]
Depending on how long it will take https://github.com/etcd-io/etcd/issues/12698 to be fixed, maybe we should merge this as is.

**Comment by thaJeztah** [comment]
> Depending on how long it will take [etcd-io/etcd#12698](https://github.com/etcd-io/etcd/issues/12698) to be fixed, maybe we should merge this as is

It was previously a CoreOS project; do you know if the maintainers are now at Red Hat? (if so, do you have some internal cages to rattle? 😇 😅).

**Comment by joakim-tjernlund** [comment]
This is not moving forward ...

**Comment by joakim-tjernlund** [comment]
Guys, this is a really old bug, was fixed many months ago. It also reflects badly on docker/moby as this bug makes life miserably for OTHER apps but docker and you don't care.
Please make sure to merge it into maintenance branches as well.

**Comment by thaJeztah** [comment]
@joakim-tjernlund not helpful; might want to complain in the upstream etcd project instead

**Comment by kolyshkin** [comment]
I think this can be merged as is now since
 - etcd-io/etcd#12698 (tag a release) is not moving forward;
 - this no longer uses my fork of etcd, but upstream repo.

**Comment by kolyshkin** [comment]
Backport to 20.10 should also be trivial (once this is merged).

**Comment by joakim-tjernlund** [comment]
> @joakim-tjernlund not helpful; might want to complain in the upstream etcd project instead

I have complained to etcd, nothing happens.
You can only wait so long for upstream. You are bundling this pkg so it becomes your problem in the end.

**Comment by joakim-tjernlund** [comment]
> Backport to 20.10 should also be trivial (once this is merged).

What etcd does 20.10.x use? 19.x is getting old and soon obsolete so 20.10 is the way forward.

**Comment by joakim-tjernlund** [comment]
Yet a release but without this fix ...

**Comment by joakim-tjernlund** [comment]
If you move to the etcd 3.4 branch you will get the fix ...
3.3 is apparently  dead since long .

**Comment by joakim-tjernlund** [comment]
3.3.27 has been out for quite a while now ...

**Comment by thaJeztah** [comment]
Rebased and moved to v3.3.27

**Review by kolyshkin — COMMENTED** [review]
LGTM (I'm the author so I can't technically LGTM it)

**Comment by kolyshkin** [comment]
I guess this needs to be backported as well.

**Review by tianon — APPROVED** [review]
LGTM; the test failure seems like a teardown race, not necessarily related?

**Review by thaJeztah — APPROVED** [review]
LGTM (although the last push was from me 😂)

**Comment by thaJeztah** [comment]
> I guess this needs to be backported as well.

I added the cherry-pick label, but it may need https://github.com/moby/moby/commit/2bef937507e77b976ffb2baf22d20c15edf24391 to be cherry-picked as well, which brings updated dependencies, so a rather large diff (perhaps we're able to reduce the diff)

**Comment by thaJeztah** [comment]
I re-ran CI a couple of times to make sure the `vendor` check also passed, but it's a bit of a whack-a-mole with flaky tests in different Jenkins stages.

The remaining failure in the last run is `TestFollowLogsHandleDecodeErr`, which is known flaky, which will be addressed by https://github.com/moby/moby/pull/43105 once merged;

```
 === Failed
 === FAIL: daemon/logger/loggerutils TestFollowLogsHandleDecodeErr (0.01s)
 time="2022-01-04T23:38:45Z" level=error msg="Log reader notified that it must re-open log file, some log data may not be streamed to the client." file=/tmp/TestFollowLogsHandleDecodeErr497049835
     logfile_test.go:317: assertion failed: 5 (int) != 6 (dec.resetCount int)
```


I'll go ahead and bring this one in

**Comment by joakim-tjernlund** [comment]
> I guess this needs to be backported as well.

Did this get backported ?

**Pull request merged**
